core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -211,6 +203,43 @@ private int readVersionHint() {
     }
   }
 
+  /**
+   * Renames the source file to destination, using the provided file system. If rename failed,
+   * the source file will be attempted to deleted.
+   *
+   * @param fs the filesystem used for the rename
+   * @param src the source file
+   * @param dst the destination file
+   */
+  private void renameToFinal(final FileSystem fs, final Path src, final Path dst) {
+    try {
+      if (!fs.rename(src, dst)) {
+        final CommitFailedException cfe = new CommitFailedException(
+            "Failed to commit changes using rename: %s", dst);
+        deleteFileNoThrow(src, cfe);
+        throw cfe;
+      }
+    } catch (IOException e) {
+      final CommitFailedException cfe = new CommitFailedException(e,
+          "Failed to commit changes using rename: %s", dst);
+      deleteFileNoThrow(src, cfe);
+      throw cfe;
+    }
+  }
+
+  /**
+   * Deletes the file from the file system. Any RuntimeException will be suppressed.
+   *
+   * @param tempMetadataFile the file to be deleted.
+   */
+  private void deleteFileNoThrow(Path tempMetadataFile, CommitFailedException cfe) {
```

Comment:
I think it would be cleaner if this caught and returned the exception instead of passing the commit exception into this method. This doesn't need to be more complicated when it can just return the RuntimeException, if there was one.

I would change it to this:
```java
private RuntimeException tryDelete(Path location) {
  try {
    io().deleteFile(location.toString);
    return null;
  } catch (RuntimeException re) {
    return re;
  }
}
```

Then the caller should check whether there is a suppressed exception and add it to the failure that will be thrown. This method doesn't need to know about that concern.

Reply:
Thanks for the discussion. I agree what you suggested is a cleaner API.

I initially thought about this too. The tradeoff here is that, in the caller (after the rename call), there will be some duplicate code on checking the return value, and `addSuppressed` if so, then throw. Extracting this part to another method like `suppressThenThrow` seems a little over to me. 😉 

Anyways, I went with what you suggested here given your preference, and left the caller as-is. Let me know what you think.

Reply:
This looks good to me now. I'll take a little duplicate code that is simple over mixing concerns in a method.